packages/common/infra/src/orm/core/table.ts

```diff
@@ -0,0 +1,155 @@
+
+type PrimaryKeyField<T extends TableSchemaBuilder> = {
+  [K in keyof T]: T[K] extends FieldSchemaBuilder<any, any, infer PrimaryKey>
+    ? PrimaryKey extends true
+      ? K
+      : never
+    : never;
+}[keyof T];
+
+export type NonPrimaryKeyFields<T extends TableSchemaBuilder> = {
+  [K in keyof T]: T[K] extends FieldSchemaBuilder<any, any, infer PrimaryKey>
+    ? PrimaryKey extends false
+      ? K
+      : never
+    : never;
+}[keyof T];
+
+export type PrimaryKeyFieldType<T extends TableSchemaBuilder> =
+  T[PrimaryKeyField<T>] extends FieldSchemaBuilder<infer Type>
+    ? Type extends Key
+      ? Type
+      : never
+    : never;
+
+export type CreateEntityInput<T extends TableSchemaBuilder> = Pretty<
+  RequiredFields<T> & OptionalFields<T>
+>;
+
+export type UpdateEntityInput<T extends TableSchemaBuilder> = Pretty<{
+  [key in NonPrimaryKeyFields<T>]?: T[key] extends FieldSchemaBuilder<
+    infer Type
+  >
+    ? Type
+    : never;
+}>;
+
+export class Table<T extends TableSchemaBuilder> {
+  readonly schema: TableSchema;
+  readonly keyField: string = '';
+  private readonly subscribedKeys: Map<Key, Subject<any>> = new Map();
+  constructor(
+    private readonly adapter: ORMAdapter,
+    public readonly name: string,
+    private readonly schemaBuilder: T
+  ) {
+    validators.validateTableSchema(this);
+
+    this.schema = Object.entries(this.schemaBuilder).reduce(
+      (acc, [fieldName, fieldBuilder]) => {
+        acc[fieldName] = fieldBuilder.schema;
+        if (fieldBuilder.schema.isPrimaryKey) {
+          // @ts-expect-error still in constructor
+          this.keyField = fieldName;
+        }
+        return acc;
+      },
+      {} as TableSchema
+    );
+  }
+
+  // @TODO(@forehalo): return value need to be specified with `Default` inference
+  async create<Input extends CreateEntityInput<T>>(
+    input: Input
+  ): Promise<Input> {
+    const data = Object.entries(this.schema).reduce(
+      (acc, [key, schema]) => {
+        const inputVal = acc[key];
+
+        if (inputVal === undefined) {
+          if (schema.optional) {
+            acc[key] = null;
+          }
+
+          if (schema.default) {
+            acc[key] = schema.default() ?? null;
+          }
+        }
+
+        return acc;
+      },
+      omitBy(input, isUndefined) as any
+    );
+
+    validators.validateCreateEntityData(this, data);
+
+    return await this.adapter.create(this.name, data[this.keyField], data);
+  }
+
+  update(
+    key: PrimaryKeyFieldType<T>,
+    input: UpdateEntityInput<T>
+  ): Promise<CreateEntityInput<T>> {
+    validators.validateUpdateEntityData(this, input);
+    return this.adapter.update(this.name, key, omitBy(input, isUndefined));
+  }
+
+  get(key: PrimaryKeyFieldType<T>) {
+    return this.adapter.read(this.name, key);
+  }
+
+  async get$(key: PrimaryKeyFieldType<T>) {
+    let subject$: Subject<any> | undefined = this.subscribedKeys.get(key);
+
+    if (!subject$) {
+      subject$ = new Subject();
```

Comment:
i think we should use BehaviorSubject

Reply:
I found using subject here is a bad idea.
there is no way to destroy the subscription